src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1644,6 +1685,25 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
             }
         }
 
+        private ImmutableArray<Result> VisitArgumentsEvaluate(
+            ImmutableArray<BoundExpression> arguments,
+            ImmutableArray<RefKind> refKindsOpt,
+            ImmutableArray<ParameterSymbol> parameters,
+            ImmutableArray<int> argsToParamsOpt,
+            bool expanded)
+        {
+            var savedState = this.State.Clone();
+            // We do a first pass to work through the arguments without making any assumptions
+            var results = VisitArgumentsEvaluate(arguments, refKindsOpt, expanded);
+
+            // We do a second pass through the arguments, ignoring any diagnostics produced, but honoring the annotations,
+            // to get the proper result state.
+            this.SetState(savedState);
```

Comment:
:memo: I suspect there is no need to reset state between the two passes, because:
- it's only possible for the second pass to produce a more restrictive state (more variables are considered non-null),
- starting from an incorrect state would affect diagnostics, but they are silenced in the second pass.